web/src/OverviewResourcePane.tsx

```diff
@@ -46,12 +48,37 @@ export default function OverviewResourcePane(props: OverviewResourcePaneProps) {
     selectedTab = r.name
   }
 
-  let alerts: Alert[] = []
-  if (r) {
-    alerts = combinedAlerts(r, null)
-  } else if (all) {
-    resources.forEach((r) => alerts.push(...combinedAlerts(r, null)))
+  const buildAlerts = () => {
+    let alerts: Alert[] = []
+    if (r) {
+      alerts = combinedAlerts(r, logStore)
+    } else if (all) {
+      resources.forEach((r) => alerts.push(...combinedAlerts(r, logStore)))
+    }
+    return alerts
   }
+  const [alerts, setAlerts] = useState<Alert[]>([])
+  // update alerts whenever the selected resource changes
+  useEffect(() => setAlerts(buildAlerts), [r])
```

Comment:
I'm pretty confident this isn't going to work like you expect. You really don't want circular dependencies between hooks like this.

instead of making alerts a hook, maybe just create a dummy state variable?

```
const [trancateCount, setTruncateCount] = useState(0)

let alerts = ...
```

it's a little janky, but it should force the alerts to regenerate correctly each time the truncate count increments, which is closer to the data flow you want.

Reply:
Hmm...not sure I totally understand how that would help here.

You're suggesting something like this?
```javascript
  const [truncateCount, setTruncateCount] = useState<number>(0)

  // add a listener to rebuild alerts whenever a truncation event occurs
  useEffect(() => {
    const rebuildAlertsOnLogClear = (e: LogUpdateEvent) => {
      if (e.action === LogUpdateAction.truncate) {
        setTruncateCount(truncateCount + 1)
      }
    }

    logStore.addUpdateListener(rebuildAlertsOnLogClear)
    return () => logStore.removeUpdateListener(rebuildAlertsOnLogClear)
  }, [])

  let alerts: Alert[] = []
  if (r) {
    alerts = combinedAlerts(r, logStore)
  } else if (all) {
    resources.forEach((r) => alerts.push(...combinedAlerts(r, logStore)))
  }
```
That won't actually force a re-render, unfortunately.

Because alerts need to react to changes in `LogStore`, they need to be in state so React can see the change and re-render. But as a result, that means that I need to handle props changes.

The "subscription" hook for `LogStore` has `[]` for deps, so it'll just subscribe on mount and unsubscribe on unmount.
The hook here is for props changes, so has `[r]` for deps, so it'll only run when the resource changes.

I believe this is correct use of hooks - setting state is generally safe as long as correctly guarded with dependencies to prevent an infinite re-render loop, and pretty common for subscription cases.

That said, one thought would be to modify the `LogStore` subscriptions (which might address some of the other feedback) so that you can pass a resource/manifest name on subscribe and only get callbacks if that one changes. Then, a single effect would suffice:
```javascript
  // add a listener to rebuild alerts whenever resource changes or a truncation event occurs
  useEffect(() => {
    const rebuildAlertsOnLogClear = (e: LogUpdateEvent) => {
      if (e.action === LogUpdateAction.truncate) {
        setAlerts(buildAlerts)
      }
    }
    
    setAlerts(buildAlerts)

    logStore.addUpdateListener(rebuildAlertsOnLogClear, r?.name)
    return () => logStore.removeUpdateListener(rebuildAlertsOnLogClear)
  }, [r])
```

Reply:
> That won't actually force a re-render, unfortunately.

you made a mistake - you'll need to create a useEffect() dependency on truncateCount, because it reads truncateCount. it should work after that.

> Because alerts need to react to changes in LogStore, they need to be in state so React can see the change and re-render.

Yep, we agree this is the core issue. But you're not handling dependencies correctly. Here's a repro case you can try yourself:

1) add `cmd='exit 1` to the storybook resource in web/storybook.tiltfile
2) Run tilt up
3) Navigate to http://localhost:10350/r/(all)/overview
4) Click 'clear logs'
5) Click the reload button on the storybook item

Expected result:
The error count should be '1'

Actual result:
The error count is '0'

We can talk about how you could handle this correctly, but i think the approach i described (where the data is unidirectional) is going to be simpler to reason about.

Reply:
D'oh - ok, everything works well with that approach (when I use dependencies properly 😝). Changes are in b8376cb, please unresolve if I did something else whacky!